cmake/Modules/CodeCoverage.cmake

```diff
@@ -337,209 +341,347 @@ function(ADD_CODE_COVERAGE)
-        ${PROJECT_BINARY_DIR}/${Coverage_NAME}.info.total
-        ${PROJECT_BINARY_DIR}/cpp_coverage/${Coverage_NAME}.info.total
-        || echo "WARNING: No cpp report to move"
+        if [ -f ${PROJECT_BINARY_DIR}/rostest_cpp_coverage/${Coverage_NAME}_rostest_cpp.info.total ]\; then
+          ${LCOV_PATH} ${LCOV_EXTRA_FLAGS} --remove
+          ${PROJECT_BINARY_DIR}/rostest_cpp_coverage/${Coverage_NAME}_rostest_cpp.info.total
+          ${LCOV_REMOVES}
+          --output-file
+          ${PROJECT_BINARY_DIR}/rostest_cpp_coverage/${Coverage_NAME}_rostest_cpp.info.removed\;
+        fi
       COMMAND
-        ${CMAKE_COMMAND} -E rename
-        ${PROJECT_BINARY_DIR}/${Coverage_NAME}.info.removed
-        ${PROJECT_BINARY_DIR}/cpp_coverage/${Coverage_NAME}.info.removed
-        || echo "WARNING: No cpp report to move"
+        if [ -f ${PROJECT_BINARY_DIR}/rostest_cpp_coverage/${Coverage_NAME}_rostest_cpp.info.removed ]\; then
+          ${LCOV_PATH} ${LCOV_EXTRA_FLAGS}
+          --extract
+          ${PROJECT_BINARY_DIR}/rostest_cpp_coverage/${Coverage_NAME}_rostest_cpp.info.removed
+          "'*${REAL_SOURCE_DIR}*'"
+          --output-file
+          ${PROJECT_BINARY_DIR}/rostest_cpp_coverage/${Coverage_NAME}_rostest_cpp.info.cleaned\;
+        fi
       COMMAND
-        ${CMAKE_COMMAND} -E rename
-        ${PROJECT_BINARY_DIR}/${Coverage_NAME}.info.cleaned
-        ${PROJECT_BINARY_DIR}/cpp_coverage/${Coverage_NAME}.info.cleaned
-        || echo "WARNING: No cpp report to move"
+        if [ -f ${PROJECT_BINARY_DIR}/rostest_cpp_coverage/${Coverage_NAME}_rostest_cpp.info.cleaned ]\; then
+          ${CMAKE_COMMAND} -E copy
+          ${PROJECT_BINARY_DIR}/rostest_cpp_coverage/${Coverage_NAME}_rostest_cpp.info.cleaned
+          ${PROJECT_BINARY_DIR}/${Coverage_NAME}_rostest_cpp.info\;
+        fi
       WORKING_DIRECTORY ${PROJECT_BINARY_DIR}
       DEPENDS
-        _run_tests_${PROJECT_NAME}
+        _run_tests_${PROJECT_NAME}_rostest
+        _create_coverage_report_${PROJECT_NAME}_rostest_cpp
+        ${PROJECT_BINARY_DIR}/rostest_cpp_coverage/${Coverage_NAME}_rostest_cpp.info.original
         _run_tests_${PROJECT_NAME}_cpp_base_coverage_report
         ${PROJECT_BINARY_DIR}/${Coverage_NAME}_base_cpp.info
     )
-    add_custom_target(${Coverage_NAME}_cpp_info
-                      DEPENDS ${PROJECT_BINARY_DIR}/${Coverage_NAME}_cpp.info)
+    add_custom_target(
+      ${Coverage_NAME}_rostest_cpp_info
+      DEPENDS
+        _run_tests_${PROJECT_NAME}_rostest
+        ${PROJECT_BINARY_DIR}/${Coverage_NAME}_rostest_cpp.info
+      )
+  else()
+    # dummy targets for the case if there is no rostest
+    add_custom_target(${Coverage_NAME}_rostest_cpp_info
+      COMMAND "${CMAKE_COMMAND}" "-E" "echo" "Skipping rostest cpp coverage report target."
+      DEPENDS _run_tests_${PROJECT_NAME})
+  endif()
+
+  # run _create_coverage_report_${PROJECT_NAME}_gtest_cpp
+  # before all rostest target tests_${PROJECT_NAME}_rostest
```

Comment:
Please add a comment about why you need to run `_create_coverage_report_${PROJECT_NAME}_gtest_cpp` first.